overlord/configstate/config/transaction.go

```diff
@@ -178,8 +226,17 @@ func (t *Transaction) Get(snapName, key string, result interface{}) error {
 		return err
 	}
 
-	// commit changes onto a copy of pristine configuration, so that get has a complete view of the config.
+	// merge virtual config and then commit changes onto a copy of pristine configuration, so that get has a complete view of the config.
 	config := t.copyPristine(snapName)
+	// 1. merge virtual
+	mergedConfig, err := mergeConfigWithVirtual(snapName, jsonRaw(config))
+	if err != nil {
+		return err
+	}
+	if err := jsonutil.DecodeWithNumber(bytes.NewReader(*mergedConfig), &config); err != nil {
+		return err
```

Comment:
I wonder about the need to decode entire config here (which we didn't do before), can mergeConfigWithVirtual not encode it, or is it too annoying / other reasons?

Reply:
Thanks, that is a great point. 

I am currently doing it this way to avoid having to change the helpers. The datatypes I have are:
1. `map[string]*json.RawMessage` from copyPristine()
2. but if I want to re-use `PatchConfig` I need to pass either *json.RawMessage or map[string]interface{}. that's why there is a jsonRaw(config) in there
3. now I have a *json:RawMessage but to `applyChanges()` I need a map[string]*json.RawMessage, this is why it is decoded again.

The original PR was changing the helpers but that was not quite loved so this version avoids it on the expense of some intermediate conversions. Happy to change of course if we prefer to change the (exiting) helpers or if there is a way I overlooked.

Reply:
Ah, I see, I had a gut feeling it was something like that. Fair enough, I think it's ok, thanks for explanation.

Reply:
Fwiw, I think we can a avoid the coding/recoding via the below diff. The downside is that there is a bit of code duplication in PatchConfig now.
```go
diff --git a/overlord/configstate/config/helpers.go b/overlord/configstate/config/helpers.go
index 44c3e9871d..036ba01f15 100644
--- a/overlord/configstate/config/helpers.go
+++ b/overlord/configstate/config/helpers.go
@@ -97,6 +97,10 @@ func PatchConfig(snapName string, subkeys []string, pos int, config interface{},
                // Raw replaces pristine on commit. Unpack, update, and repack.
                var configm map[string]interface{}
 
+               if config == nil {
+                       config = jsonRaw(nil)
+               }
+
                if err := jsonutil.DecodeWithNumber(bytes.NewReader(*config), &configm); err != nil {
                        return nil, fmt.Errorf("snap %q option %q is not a map", snapName, strings.Join(subkeys[:pos], "."))
                }
@@ -120,6 +124,20 @@ func PatchConfig(snapName string, subkeys []string, pos int, config interface{},
                // support cases where a previously unset path is set back.
                return jsonRaw(result), nil
 
+       case map[string]*json.RawMessage:
+               // Update map to apply against pristine on commit.
+               if pos+1 == len(subkeys) {
+                       config[subkeys[pos]] = value
+                       return config, nil
+               } else {
+                       result, err := PatchConfig(snapName, subkeys, pos+1, config[subkeys[pos]], value)
+                       if err != nil {
+                               return nil, err
+                       }
+                       config[subkeys[pos]] = jsonRaw(result)
+                       return config, nil
+               }
+
        case map[string]interface{}:
                // Update map to apply against pristine on commit.
                if pos+1 == len(subkeys) {
diff --git a/overlord/configstate/config/transaction.go b/overlord/configstate/config/transaction.go
index e70bc03d2a..9decbd94dd 100644
--- a/overlord/configstate/config/transaction.go
+++ b/overlord/configstate/config/transaction.go
@@ -227,11 +227,7 @@ func (t *Transaction) Get(snapName, key string, result interface{}) error {
        // merge virtual config and then commit changes onto a copy of pristine configuration, so that get has a complete view of the config.
        config := t.copyPristine(snapName)
        // 1. merge virtual
-       mergedConfig, err := mergeConfigWithVirtual(snapName, jsonRaw(config))
-       if err != nil {
-               return err
-       }
-       if err := jsonutil.DecodeWithNumber(bytes.NewReader(*mergedConfig), &config); err != nil {
+       if err := mergeConfigWithVirtual(snapName, config); err != nil {
                return err
        }
        // 2. apply changes
@@ -406,12 +402,12 @@ func commitChange(pristine *json.RawMessage, change interface{}) *json.RawMessag
 // with the virtual configuration values by calling the registered
 // virtual configuration function of the given snap. The merged config
 // is returned.
-func mergeConfigWithVirtual(instanceName string, config *json.RawMessage) (*json.RawMessage, error) {
+func mergeConfigWithVirtual(instanceName string, config map[string]*json.RawMessage) error {
        virtualMu.Lock()
        km, ok := virtualMap[instanceName]
        virtualMu.Unlock()
        if !ok {
-               return config, nil
+               return nil
        }
 
        // create a "patch" from the virtual entries
@@ -422,7 +418,7 @@ func mergeConfigWithVirtual(instanceName string, config *json.RawMessage) (*json
                }
                res, err := virtualFn(instanceName, key)
                if err != nil {
-                       return nil, err
+                       return err
                }
                patch[key] = jsonRaw(res)
        }
@@ -431,14 +427,12 @@ func mergeConfigWithVirtual(instanceName string, config *json.RawMessage) (*json
        patchKeys := sortPatchKeysByDepth(patch)
        for _, subkeys := range patchKeys {
                raw := jsonRaw(patch[subkeys])
-               mergedConfig, err := PatchConfig(instanceName, strings.Split(subkeys, "."), 0, config, raw)
-               if err != nil {
-                       return nil, err
+               if _, err := PatchConfig(instanceName, strings.Split(subkeys, "."), 0, config, raw); err != nil {
+                       return err
                }
-               config = jsonRaw(mergedConfig)
        }
 
-       return config, nil
+       return nil
 }
 
 // clearVirtualConfig iterates over a given config and removes any values
```